File: content/snippets/python/exercises/advanced/graph_traversal.py
from collections import deque

g = {0:[1,2],1:[0,3],2:[0,3],3:[1,2]}
print('BFS:', end=' ')
q=deque([0]); seen={0}
while q:
    u=q.popleft(); print(u, end=' ')
    for v in g[u]:
        if v not in seen: seen.add(v); q.append(v)
print('\nDFS:', end=' ')
seen=set()
import sys
sys.setrecursionlimit(10000)
def dfs(u):
    seen.add(u); print(u, end=' ')
    for v in g[u]:
        if v not in seen: dfs(v)

dfs(0)

File: content/snippets/go/exercises/advanced/graph_traversal.go
package main

import "fmt"

func main() {
	g := [][]int{{1, 2}, {0, 3}, {0, 3}, {1, 2}}
	// BFS
	q := []int{0}
	seen := map[int]bool{0: true}
	fmt.Print("BFS:")
	for len(q) > 0 {
		u := q[0]
		q = q[1:]
		fmt.Print(" ", u)
		for _, v := range g[u] {
			if !seen[v] {
				seen[v] = true
				q = append(q, v)
			}
		}
	}
	fmt.Println()
	// DFS
	fmt.Print("DFS: ")
	seen = map[int]bool{}
	var dfs func(int)
	dfs = func(u int) {
		seen[u] = true
		fmt.Print(u, " ")
		for _, v := range g[u] {
			if !seen[v] {
				dfs(v)
			}
		}
	}
	dfs(0)
	fmt.Println()
}
